src/api/parser/src/data/feed.js

```diff
@@ -0,0 +1,202 @@
+const normalizeUrl = require('normalize-url');
+const { hash } = require('@senecacdot/satellite');
+
+const {
+  getFeed,
+  getFeeds,
+  addFeed,
+  removeFeed,
+  setInvalidFeed,
+  isInvalid,
+  setDelayedFeed,
+  isDelayed,
+  removePost,
+  getPost,
+  getPosts,
+  getFlaggedFeeds,
+  setFlaggedFeed,
+  unsetFlaggedFeed,
+} = require('../storage');
+
+const urlToId = (url) => hash(normalizeUrl(url));
+
+class Feed {
+  constructor(author, url, user, link, etag, lastModified) {
+    if (!url) {
+      throw new Error('missing url for feed');
+    }
+    if (!author) {
+      throw new Error('missing author for feed');
+    }
+    // Use the feed's normalized url as our unique identifier
+    this.id = urlToId(url);
+    this.author = author;
+    this.url = url;
+    this.user = user;
+    this.link = link;
+
+    // We may or may not have these cache values when we create a feed.
+    this.etag = etag === '' ? null : etag;
+    this.lastModified = lastModified === '' ? null : lastModified;
+  }
+
+  /**
+   * Save the current Feed to the database.
+   * Returns a Promise.
+   */
+  save() {
+    return addFeed(this);
+  }
+
+  /**
+   * Removes current Feed + associated Posts from the databases.
+   * Returns a Promise
+   */
+  async delete() {
+    // Removing feeds and getting all posts, we'll be assigning all the Posts we get back to posts
+    let [, posts] = await Promise.all([removeFeed(this.id), getPosts(0, 0)]);
+
+    // Filter out all posts which do not contain feed id of feed being removed
+    posts = await Promise.all(posts.map((id) => getPost(id)));
+    posts = posts.filter((post) => post.feed === this.id);
+
+    // Remove the post from Redis + ElasticSearch
+    await Promise.all([].concat(posts.map((post) => removePost(post.id))));
+  }
+
+  /**
+   * Updates current Feed in the database.
+   * Returns a Promise.
+   */
+  async update() {
+    await this.delete();
+    await this.save();
+  }
+
+  /**
+   * Adds the current Feed to the database with the specified reason
+   * Returns a Promise
+   */
+  setInvalid(reason) {
+    return setInvalidFeed(this.id, reason);
+  }
+
+  /**
+   * Checks whether the current feed is valid or not
+   * Returns a Promise<Boolean>.
+   */
+  isInvalid() {
+    return isInvalid(this.id);
+  }
+
+  /**
+   * Flags a feed in the database, indicating that its processing should be delayed
+   * @param {Number} seconds - duration in seconds for which processing should wait
+   * Returns a Promise
+   */
+  setDelayed(seconds) {
+    return setDelayedFeed(this.id, seconds);
+  }
+
+  /**
+   * Checks whether the current feed is delayed or not
+   * Returns a Promise<Boolean>
+   */
+  async isDelayed() {
+    return (await isDelayed(this.id)) === '1';
+  }
+
+  /**
+   * Flags the feed, preventing posts from the feed to be displayed
+   * Returns a Promise<Boolean>
+   */
+  flag() {
```

Comment:
This stuff will probably need to be changed later to use what's in the Users data vs. what's in Redis.  For now, this is fine, but FYI so we don't forget.